File: virtual-world-cpp/VirtualWorld/World.cpp
#include "World.h"
#include <iostream>
#include <algorithm>
#include <queue>
#include "AllOgranisms.h"

World::World(int width, int height) : width(width), height(height) {
    map.resize(width * height);
}

World::~World() {
    Empty();
}

void World::Print(int x, int y, int range) {
    int startX = x - range;
    int startY = y - range;
    if(startX < 0)
        startX = 0;
    if(startY < 0)
        startY = 0;
    int endX = x + range;
    int endY = y + range;
    if(startX == 0)
        endX = range * 2;
    if(startY == 0)
        endY = range * 2;
    if(endX >= width)
        startX = width - range * 2 - 1;
    if(endY >= height)
        startY = height - range * 2 - 1;
    if(endX >= width)
        endX = width - 1;
    if(endY >= height)
        endY = height - 1;
    if(startX < 0)
        startX = 0;
    if(startY < 0)
        startY = 0;
    std::cout << "\xDB";
    for(int i = startX; i <= endX; i++)
        std::cout << "\xDF\xDF";
    std::cout << "\xDB\n";
    for (int itY = startY; itY <= endY; itY++) {
        std::cout << "\xDB";
        for (int itX = startX; itX <= endX; itX++) {
            if (IsOccupied(itX, itY))
                std::cout << GetOrganism(itX, itY)->getSymbol() << " ";
            else
                std::cout << "  ";
        }
        std::cout << "\xDB";
        if(itY - startY < messages.size())
            std::cout << "\t" << messages[itY - startY];
        std::cout << std::endl;
    }
    std::cout << "\xDB";
    for(int i = startX; i <= endX; i++)
        std::cout << "\xDC\xDC";
    std::cout << "\xDB";
}

Tile *World::GetTile(int x, int y) {
    if(x < 0 || x >= width || y < 0 || y >= height)
        throw std::runtime_error("Tile out of bounds");
    return &map[y * width + x];
}

Organism *World::GetOrganism(int x, int y) {
    return GetTile(x, y)->GetOrganism();
}

bool World::IsOccupied(int x, int y) {
    return GetTile(x, y)->IsOccupied();
}

void World::AddOrganism(Organism *organism) {
    organisms.push_back(organism);
    Tile *tile = GetTile(organism->getX(), organism->getY());
    if(tile->IsOccupied())
        throw std::runtime_error("Tile is already occupied");
    tile->SetOrganism(organism);
}

int World::GetWidth() {
    return width;
}

int World::GetHeight() {
    return height;
}

void World::MoveOrganism(Organism *organism, int x, int y) {
    if(x < 0 || x >= width || y < 0 || y >= height)
        throw std::runtime_error("Invalid coordinates");
    Tile *tile = GetTile(x, y);
    if(tile->IsOccupied())
        throw std::runtime_error("Tile is already occupied");
    Tile *oldTile = GetTile(organism->getX(), organism->getY());
    oldTile->SetOrganism(nullptr);
    tile->SetOrganism(organism);
    organism->setX(x);
    organism->setY(y);
}

void World::Update() {
    struct OrganismComparator {
        bool operator()(Organism *a, Organism *b) {
            if(a->getInitiative() == b->getInitiative())
                return a->getAge() < b->getAge();
            return a->getInitiative() < b->getInitiative();
        }
    };
    std::priority_queue<Organism*, std::vector<Organism*>, OrganismComparator> queue;
    for(Organism *organism : organisms)
        queue.push(organism);
    while(!queue.empty()) {
        Organism *organism = queue.top();
        queue.pop();
        if(!organism->IsAlive())
            continue;
        organism->Action();
        organism->setAge(organism->getAge() + 1);
    }
}

void World::Randomize() {
    static const int spawnRate[] = {3, 4, 10, 4, 4, 4, 2, 1, 1, 1, 1};
    for(int type = 0; type < typeCount; type++) {
        int count = 2 + log(width * height) * 0.005;
        count *= spawnRate[type];
        for(int i = 0; i < count; i++) {
            int x = rand() % width;
            int y = rand() % height;
            if (!IsOccupied(x, y)) {
                switch (type) {
                    case 0:
                        AddOrganism(new Wolf(x, y, this));
                        break;
                    case 1:
                        AddOrganism(new Sheep(x, y, this));
                        break;
                    case 2:
                        AddOrganism(new Grass(x, y, this));
                        break;
                    case 3:
                        AddOrganism(new Fox(x, y, this));
                        break;
                    case 4:
                        AddOrganism(new Turtle(x, y, this));
                        break;
                    case 5:
                        AddOrganism(new Antelope(x, y, this));
                        break;
                    case 6:
                        AddOrganism(new Dandelion(x, y, this));
                        break;
                    case 7:
                        AddOrganism(new Guarana(x, y, this));
                        break;
                    case 8:
                        AddOrganism(new Nightshade(x, y, this));
                        break;
                    case 9:
                        AddOrganism(new SosnowskysHogweed(x, y, this));
                        break;
                    case 10:
                        AddOrganism(new CyberSheep(x, y, this));
                        break;
                }
            }
        }
    }
}

void World::RemoveOrganism(Organism *organism) {
    GetTile(organism->getX(), organism->getY())->SetOrganism(nullptr);
    organism->Kill();
}

void World::Empty() {
    for(Organism* organism : organisms)
        delete organism;
    organisms.clear();
}

void World::RemoveDead() {
    for(int i = 0; i < organisms.size(); i++) {
        if(!organisms[i]->IsAlive()) {
            Tile* tile = GetTile(organisms[i]->getX(), organisms[i]->getY());
            if(tile->GetOrganism() == organisms[i])
                tile->SetOrganism(nullptr);
            delete organisms[i];
            organisms.erase(organisms.begin() + i);
            i--;
        }
    }
}

std::vector<std::string> &World::GetMessages() {
    return messages;
}

void World::AddMessage(const std::string& message) {
    messages.push_back(message);
}

void World::ClearMessages() {
    messages.clear();
}

std::vector<Organism *> &World::GetOrganisms() {
    return organisms;
}

int World::GetOrganismCount() {
    return organisms.size();
}

File: virtual-world-java/src/main/java/me/knasuta/virtualworld/simulation/World.java
package me.knasuta.virtualworld.simulation;

import me.knasuta.virtualworld.simulation.animals.Human;

import java.io.*;
import java.util.PriorityQueue;
import java.util.Vector;

public class World implements Serializable {
    private final int width;
    private final int height;
    private final Vector<Organism> organisms;
    private final Vector<String> messages;
    private final Organism[][] organismMap;
    private int turn;
    private final boolean hexagonal;
    private Human player;

    public Human getPlayer(){
        return player;
    }
    public void Update(){
        this.turn++;
        this.ClearMessages();
        PriorityQueue<Organism> organisms = new PriorityQueue<>(this.organisms);
        while(!organisms.isEmpty()){
            Organism organism = organisms.poll();
            if(organism.IsAlive())
                organism.Action();
            organism.setAge(organism.getAge()+1);
        }
        this.RemoveDeadOrganisms();
    }

    public World(int width, int height, boolean hexagonal) {
        this.width = width;
        this.height = height;
        this.organisms = new Vector<Organism>();
        this.messages = new Vector<String>();
        this.organismMap = new Organism[width][height];
        this.hexagonal = hexagonal;
        this.turn = 0;
    }
    public World(int width, int height) {
        this(width, height, false);
    }
    public Organism getOrganism(int x, int y) throws IndexOutOfBoundsException {
        if(x < 0 || x >= this.width || y < 0 || y >= this.height) {
            throw new IndexOutOfBoundsException("Cannot get organism outside of world");
        }
        return this.organismMap[x][y];
    }
    public Organism getOrganism(Point point){
        return getOrganism(point.getX(), point.getY());
    }
    public boolean IsOccupied(int x, int y) {
        return this.organismMap[x][y] != null;
    }
    public void AddOrganism(Organism organism) {
        if(this.IsOccupied(organism.getX(), organism.getY())) {
            return;
        }
        this.organisms.add(organism);
        this.organismMap[organism.getX()][organism.getY()] = organism;
    }
    public void RemoveOrganism(Organism organism) {
        organism.Kill();
        this.organisms.remove(organism);
        this.organismMap[organism.getX()][organism.getY()] = null;
    }
    public void MoveOrganism(Organism organism, int x, int y)  {
        if(this.IsOccupied(x, y)) {
            return;
        }
        this.organismMap[organism.getX()][organism.getY()] = null;
        this.organismMap[x][y] = organism;
        organism.setX(x);
        organism.setY(y);
    }
    public void MoveOrganism(Organism organism, Point point)  {
        this.MoveOrganism(organism, point.getX(), point.getY());
    }
    public int getWidth() {
        return this.width;
    }
    public int getHeight() {
        return this.height;
    }
    public int getTurn() {
        return this.turn;
    }
    public boolean IsHexagonal() {
        return this.hexagonal;
    }
    public void AddMessage(String message) {
        this.messages.add(message);
    }
    public Vector<String> getMessages() {
        return this.messages;
    }
    public void ClearMessages() {
        this.messages.clear();
    }
    public void RemoveDeadOrganisms() {
        for(int i = 0; i < this.organisms.size(); i++) {
            if(!this.organisms.get(i).IsAlive()) {
                organisms.remove(i);
                i--;
            }
        }
    }
    public Vector<Point> GetAdjacentPoints(Point location){
        Vector<Point> adjacentPoints = new Vector<Point>();
        for(int x = location.getX() - 1; x <= location.getX() + 1; x++) {
            for(int y = location.getY() - 1; y <= location.getY() + 1; y++) {
                if(x >= 0 && x < this.width && y >= 0 && y < this.height && (x != location.getX() || y != location.getY())) {
                    if(this.hexagonal && (x == location.getX() - 1 && y == location.getY() - 1 || x == location.getX() + 1 && y == location.getY() + 1))
                        continue;
                    adjacentPoints.add(new Point(x, y));
                }
            }
        }
        return adjacentPoints;
    }

    public Point GetRandomAdjacentPoint(Point location) {
        Vector<Point> adjacentPoints = this.GetAdjacentPoints(location);
        if(adjacentPoints.size() == 0) {
            return null;
        }
        return adjacentPoints.get((int)(Math.random() * adjacentPoints.size()));
    }

    public void AddStartingOrganisms(){
        int x = (int)(Math.random() * this.width);
        int y = (int)(Math.random() * this.height);
        this.player = new Human(new Point(x, y), this);
        this.AddOrganism(this.player);
        int[] organismCounts = new int[OrganismType.values().length];
        organismCounts[OrganismType.WOLF.ordinal()] = 5;
        organismCounts[OrganismType.SHEEP.ordinal()] = 7;
        organismCounts[OrganismType.GRASS.ordinal()] = 4;
        organismCounts[OrganismType.ANTELOPE.ordinal()] = 7;
        organismCounts[OrganismType.FOX.ordinal()] = 6;
        organismCounts[OrganismType.TURTLE.ordinal()] = 5;
        organismCounts[OrganismType.DANDELION.ordinal()] = 2;
        organismCounts[OrganismType.GUARANA.ordinal()] = 1;
        organismCounts[OrganismType.NIGHTSHADE.ordinal()] = 1;
        organismCounts[OrganismType.SOSNOWSKY_HOGWEED.ordinal()] = 1;
        int baseOrganismCount = (int)(2 + Math.log(width * height) * 0.005f);
        IOrganismFactory factory = OrganismFactory.getInstance();
        for(OrganismType organismType : OrganismType.values()) {
            for(int i = 0; i < organismCounts[organismType.ordinal()]*baseOrganismCount; i++) {
                x = (int)(Math.random() * this.width);
                y = (int)(Math.random() * this.height);
                this.AddOrganism(factory.Create(organismType, new Point(x, y), this));
            }
        }
    }

    public Vector<Organism> getOrganisms() {
        return this.organisms;
    }

    public boolean Save(String absolutePath) {
        try {
            FileOutputStream fileOut = new FileOutputStream(absolutePath);
            ObjectOutputStream out = new ObjectOutputStream(fileOut);
            out.writeObject(this);
            out.close();
            fileOut.close();
        } catch (FileNotFoundException e) {
            return false;
        }
        catch(IOException e) {
            throw new RuntimeException(e);
        }
        return true;
    }
    public static World Load(String absolutePath) {
        try {
            FileInputStream fileIn = new FileInputStream(absolutePath);
            ObjectInputStream in = new ObjectInputStream(fileIn);
            World world = (World) in.readObject();
            in.close();
            fileIn.close();
            return world;
        } catch (FileNotFoundException e) {
            return null;
        } catch (IOException | ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
}
